src/sync_session.cpp

```diff
@@ -421,31 +448,54 @@ void SyncSession::unregister(std::unique_lock<std::mutex>& lock)
 void SyncSession::wait_for_upload_completion(std::function<void()> callback)
 {
     REALM_ASSERT(shared_from_this());
-    auto thread = std::thread([this, callback=std::move(callback), self=shared_from_this()]() {
-        {
-            std::unique_lock<std::mutex> lock(m_state_mutex);
-            if (m_session) {
-                m_session->wait_for_upload_complete_or_client_stopped();
+    auto thread = std::thread([this, callback=std::move(callback)]() {
+        std::unique_lock<std::mutex> lock(m_state_mutex);
+        if (m_state != &State::active && m_state != &State::dying) {
+            lock.unlock();
+            callback();
+        } else {
+            ++m_pending_upload_threads;
+            lock.unlock();
+            // Wait on the session and perform the user callback.
+            m_session->wait_for_upload_complete_or_client_stopped();
+            callback();
+            // Once callback is finished, if necessary perform cleanup work and kill session.
+            lock.lock();
+            --m_pending_upload_threads;
+            if (m_state == &State::dying
+                && m_pending_upload_threads == 0
+                && m_pending_download_threads == 0) {
+                // Advance state to 'inactive'
+                advance_state(lock, State::inactive);
             }
+            lock.unlock();
         }
-
-        callback();
     });
     thread.detach();
 }
 
 void SyncSession::wait_for_download_completion(std::function<void()> callback)
 {
     REALM_ASSERT(shared_from_this());
-    auto thread = std::thread([this, callback=std::move(callback), self=shared_from_this()]() {
-        {
-            std::unique_lock<std::mutex> lock(m_state_mutex);
-            if (m_session) {
-                m_session->wait_for_download_complete_or_client_stopped();
+    auto thread = std::thread([this, callback=std::move(callback)]() {
```

Comment:
There's a lot of nontrivial logic here that's entirely duplicated from the upload version.